### PR TITLE
#20903 Migrating the prerender io to dotCMS core

### LINKED ISSUE #20903: Create prerender.io App
We need to create an app that implements the prerender.io filter as a WebIntercepter and allows a user to configure  prerender and the options available 

Here is the filter
https://github.com/greengerong/prerender-java

and here are the configuration options available:
https://github.com/greengerong/prerender-java/blob/master/src/main/java/com/github/greengerong/PreRenderSEOFilter.java#L18

The initial implementation should be a plugin that can with with 5.3.8.x and we should also build this into the core

### PR DESCRIPTION
**Pull request by jdotcms**
Migrating the plugin prerender io to core

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/concurrent/ConditionalSubmitterImpl.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    @Override
+    public <T> T submit(final Supplier<T> onAvailableSupplier, final Supplier<T> onDefaultSupplier) {
+
+        T result = null;
+        if (this.semaphore.availablePermits() > 0 && this.semaphore.tryAcquire()) {
+
+            result = onAvailableSupplier.get();
+            this.semaphore.release();
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotcms/concurrent/ConditionalSubmitterImpl.java:23]: Should this be in a finally block?

**Review by nollymar — CHANGES_REQUESTED** [review]
Please check out the tests, I saw one broken related to these changes

**Review by nollymar — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/prerender/PrerenderConfig.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    public String getProtocol() {
+
+        return config.protocol;
+    }
+
+    public List<String> getExtensionsToIgnore() {
+
+        final List<String> extensionsToIgnore = Lists.newArrayList(".js", ".json", ".css", ".xml", ".less", ".png", ".jpg",
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotcms/prerender/PrerenderConfig.java:77]: Should we also read them via config prop?

File: dotCMS/src/main/java/com/dotcms/prerender/PrerenderConfig.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        if (isNotBlank(extensionsToIgnoreFromConfig)) {
+            extensionsToIgnore.addAll(Arrays.asList(extensionsToIgnoreFromConfig.trim().split(",")));
+        }
+
+        return extensionsToIgnore;
+    }
+
+    public List<String> getCrawlerUserAgents() {
+
+        final List<String> crawlerUserAgents = Lists.newArrayList("baiduspider",
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotcms/prerender/PrerenderConfig.java:91]: And read these from a config prop too?

**Review by jcastro-dotcms — CHANGES_REQUESTED** [review]

File: dotCMS/src/main/java/com/dotcms/prerender/PrerenderConfig.java (hunk 2 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/prerender/PrerenderConfig.java:77]: Nope, b.c it is handle via app config.

**Review by jcastro-dotcms — APPROVED** [review]

**Pull request merged**